src/Engine/EmitMSIL/Replication.cs

```diff
@@ -319,43 +355,180 @@ private static List<List<ReplicationGuide>> ConstructRepGuides(string[][] replic
+            //
+            // All 3 cases will run through the same matching steps.
+
+            // A static function call has an invalid this pointer and a valid class scope;
+            // TODO_MSIL: figure out the this ptr context
+            bool isValidStaticFuncCall = /*!thisPtr.IsNull && !thisPtr.IsPointer &&*/ classId != Constants.kInvalidIndex;
+            if (isValidStaticFuncCall)
+            {
+                return feps.FirstOrDefault();
+            }
+
+            // TODO_MSIL: figure out the this ptr context
+            int typeID = /*thisPtr.TypeUID;*/Constants.kInvalidIndex;
+
+            // Try to match with feps belonging to the class scope (most derived class should have priority).
+            // In this case we simply select the function that belongs to the calling class.
+            // The assumption here is that all function end points in "feps" have already been checked that they have the same signature.
+            IEnumerable<CLRFunctionEndPoint> exactFeps = feps.Where(x => x.procedureNode.ClassID == typeID);
+            if (exactFeps.Count() == 1)
+            {
+                return exactFeps.First();
+            }
+
+            //Walk the class tree structure to find the method
+            while (runtimeCore.ClassTable.ClassNodes[typeID].Base != Constants.kInvalidIndex)
+            {
+                typeID = runtimeCore.ClassTable.ClassNodes[typeID].Base;
+
+                foreach (var fep in feps)
+                    if (fep.procedureNode.ClassID == typeID)
+                        return fep;
+            }
+
+            //We weren't able to distinguish based on class hiearchy, try to sepearete based on array ranking
+            List<int> numberOfArbitraryRanks = new List<int>();
+
+            foreach (var fep in feps)
+            {
+                int numArbitraryRanks = 0;
+
+                for (int i = 0; i < argumentsList.Count; i++)
+                {
+                    if (fep.FormalParams[i].Rank == Constants.kArbitraryRank)
+                        numArbitraryRanks++;
+                }
+
+                numberOfArbitraryRanks.Add(numArbitraryRanks);
+            }
+
+            int smallest = Int32.MaxValue;
+            List<int> indicesOfSmallest = new List<int>();
+
+            for (int i = 0; i < feps.Count; i++)
+            {
+                if (numberOfArbitraryRanks[i] < smallest)
+                {
+                    smallest = numberOfArbitraryRanks[i];
+                    indicesOfSmallest.Clear();
+                    indicesOfSmallest.Add(i);
+                }
+                else if (numberOfArbitraryRanks[i] == smallest)
+                    indicesOfSmallest.Add(i);
+            }
+
+            Validity.Assert(indicesOfSmallest.Count > 0,
+                            "Couldn't find a fep when there should have been multiple: {EB589F55-F36B-404A-91DC-8D0EDC527E72}");
+
+            if (indicesOfSmallest.Count == 1)
```

Comment:
so we are looking for the FEP with the least number of arbitrary rank params?

Reply:
yes, that is what it looks like to me too